Refresh: Add Microk8s walkthrough to Edge tutorial (#49)
crate/cloud-getting-started: docs/edge.rst

.. image:: _assets/img/cloud-edge-delete.png
   :alt: CrateDB Edge deletion confirmation screen

>     :scale: 20%

Author: This bit does not work, but is unimportant - @amotl and I will work on this later. You can ignore it

Reviewer: Hi Matthijs,

thank you for tagging me on this conversation. This is a classic off-by-one error. When aligning the indentation of `:scale:` to match the first property `:alt:`, i.e. by removing a single space character, it starts working on my machine.

```suggestion
   :scale: 60%
```

With kind regards,
Andreas.

Author: Well there you go :D Thanks!

Deleting a custom region does not delete the resources inside that region. To
also delete the resources inside the region, run the script provided in the
deletion confirmation screen in your local CLI. This will uninstall CrateDB
Cloud Edge from your local Kubernetes cluster.

.. _edge-tools:

> Install CrateDB Edge with third-party tools

Reviewer 2: Do we want to call them 'tools'? These are more like Kubernetes distributions... Perhaps "Distribution-specific tutorials"?

Reviewer 3: ```suggestion
Install CrateDB Edge using a on-premise k8s distribution.
```

===========================================

In the tutorial walkthroughs below, you can read how to install CrateDB Edge
using two of the most common Kubernetes distributions: `Microk8s`_ and `K3s`_.
These are third-party tools and not officially supported by Crate.io, nor are
we responsible for their behavior. That said, we have tested the instructions
provided below for functionality. Users less familiar with customizing their
Kubernetes stack on their own may find either of these two guides a practical
solution for easier CrateDB Edge setup.

.. _edge-tools-microk8s:

Microk8s
--------

Below is a full walkthrough of the steps required to get CrateDB Edge up and
running on Microk8s. The steps are merely examples of a process validated by
us, but we do not guarantee its use, nor is it the only possible method for

> CrateDB Edge on Microk8s. As mentioned, we do not officially support Microk8s

Reviewer 2: Do we want to say that we don't officially support it? What do we support then?..

Reviewer 3: It is probably not microk8s itself, that might lead to troubles. It is what is built around it (hardware, networking, storage, etc). 

```suggestion
Below is a full walkthrough of the steps required to get CrateDB Edge up and running on Microk8s. The steps are merely examples of a process validated by us, to spin up CrateDB Edge on Microk8s. 
```

Author: What does 'it' mean here? I do think we should be quite explicit about what we support and what we do not. Having validated a certain method is also not the same as support from an SLA point of view, right?

Reviewer 2: I meant microk8s by "it". When I read "we don't officially support it" I hear "use at your own risk" and "not production ready at all". In reality we support (or don't support) it as much as Azure AKS - we've tried and it works on both, but can't guarantee everything for you?

or other third-party tools. We provide this information 'as is', for ease of
use and to illustrate the workings of CrateDB Edge.

Set up Microk8s
'''''''''''''''

> Follow the instructions from the `Microk8s docs`_:

Reviewer 2: I would also add that "In this tutorial, we'll be setting up a three node Kubernetes cluster. You can also use a single node for testing purposes."

> .. code-block:: console

Reviewer 2: Also worth adding that you need to run this on each of the microk8s nodes that you'll be clustering together.

Author: What does 'this' mean here? The whole thing, or just that section?

Reviewer 2: The entire section needs to be run on each instance you're installing microk8s on.

>     sudo snap install microk8s --classic --channel=1.21

Reviewer 2: ```suggestion
    # These instructions assume a snap-enabled distribution, such as Ubuntu or related.
    sudo snap install microk8s --classic --channel=1.21
```

    sudo usermod -a -G microk8s $USER
    sudo chown -f -R $USER ~/.kube

    microk8s status --wait-ready
    microk8s kubectl get nodes

    alias kubectl='microk8s kubectl'

    microk8s enable dns storage

Set up DNS
''''''''''

Here you set up the DNS:

.. code-block:: console

>     68.183.5.88 ub12

Reviewer 2: I think @WalBeh created these `/etc/hosts` entries for his own benefit to not have to constantly type IP addresses. I think it would be useful to clarify this, i.e.:
```
To simplify cluster bootstrapping, you can add each of the nodes IP addresses into your `/etc/hosts`, i.e.:
...
```

Reviewer 2: Or, to be fair, maybe remove the `Set up DNS` section altogether as its a matter of personal preference (and we don't refer to these DNS names in subsequent instructions)

Reviewer 3: > Or, to be fair, maybe remove the `Set up DNS` section altogether as its a matter of personal preference (and we don't refer to these DNS names in subsequent instructions)

AFAIR the nodes need to be able to resolve against "each other".

Reviewer 2: I thought you can just use IPs that `microk8s add-node` gives you?
```
❯ microk8s add-node
From the node you wish to join to this cluster, run the following:
microk8s join 192.168.1.192:25000/0d210786c86d17a78937cf64d176bdb6
```

    178.62.239.7 ub11
    188.166.9.90 ub10

Set up cluster
''''''''''''''

> Next is the command for the three-node cluster:

Reviewer 2: ```suggestion
On one of the nodes, run the command to get joining instructions. This will print the command that you need to run on the other two nodes to create a Kubernetes cluster.
```

.. code-block:: console

    microk8s add-node

Join nodes to cluster
'''''''''''''''''''''

> Now you join the nodes to the cluster:

Reviewer 2: ```suggestion
Now ssh into the two remaining nodes and run the command you received from `microk8s add-node` on the first node.
```

.. code-block:: console

>     root@ub11:~# microk8s join 188.166.9.90:25000/<cluster id>

Reviewer 2: ```suggestion
    root@ub11:~# microk8s join <ip-of-first-node>:25000/<cluster id>
```

>     Contacting cluster at 188.166.9.90

Reviewer 2: ```suggestion
    Contacting cluster at <ip-address>
```

    Waiting for this node to finish joining the cluster...

Use a storage solution
''''''''''''''''''''''

The Microk8s setup will require a storage solution. In this case, the tutorial
shows how to do so using `Longhorn`_, a distributed storage solution for
Kubernetes. You can follow the `Longhorn installation instructions`_ as
described below. (Other storage solutions for Kubernetes may work as well.)

First the installation:

.. code-block:: console

    kubectl apply -f https://raw.githubusercontent.com/longhorn/longhorn/v1.1.1/deploy/longhorn.yaml

Then you need to specify the root directory:

.. code-block:: console

>     k8 -n longhorn-system edit deployment longhorn-driver-deployer

Reviewer 2: ```suggestion
    kubectl -n longhorn-system edit deployment longhorn-driver-deployer
```

    - name: KUBELET_ROOT_DIR
    value: /var/snap/microk8s/common/var/lib/kubelet

Set up Cloud region
'''''''''''''''''''

At this stage, you can create an Edge region via the CrateDB Cloud Console.
Follow the steps outlined above :ref:`from the CrateDB sign up onwards
<edge-signup>` to proceed.

Run the script
''''''''''''''

Run the script with the following command:

.. code-block:: console

    wget -qO- https://console.cratedb.cloud/edge/cratedb-cloud-edge.sh > edge-installer.sh
    chmod u+x edge-installer.sh
    ./edge-installer --dry-run  <token>

Note that ``dry-run`` provides, as the name suggests, a method to test the

> installation by running the YAML files that are going to be applied. This gives

Reviewer 2: ```suggestion
installation by generating but not applying the Kubernetes manifests for your CrateDB Edge region. This gives
```

you an opportunity to verify them before the full install.

The ``<token>`` in question is the token you receive from the CrateDB Console
Edge region field in the Regions tab of the Organization Overview. For more
information on this section of the CrateDB Console, refer to our :ref:`CrateDB
Cloud Console overview <cloud-reference:overview-org-regions>`.

With this, you should be ready to use CrateDB Edge via Microk8s.

Microk8s Dashboard
''''''''''''''''''

With the Microk8s setup, you can optionally use the Microk8s Dashboard. To
install this dashboard, run the following in the CLI:

.. code-block:: console

    microk8s enable dashboard
    microk8s kubectl edit -n kube-system service/kubernetes-dashboard (change PORT to 8443)

>     ssh -i ~/.ssh/id_ed25519 root@<IP address> -L 10443:localhost:10443

Reviewer 2: `8443` does not match `10443`, maybe @WalBeh can clarify?

    token=$(microk8s kubectl -n kube-system get secret | grep default-token | cut -d " " -f1)

>     microk8s kubectl -n kube-system describe secret $token

Reviewer 2: These instructions didn't actually work for me... I can see the dashboard is there, but no clue how to access it.
